### PR TITLE
waf lock utility integration with Modsec

### PR DESCRIPTION
**Pull request by azhao155**
It's based on the pr https://github.com/Microsoft/ModSecurity/pull/33
and integrate with Modsec.

Testing: 
1. start nginx, send traffic. create new lock
2. reload nginx, send traffic, still use the old lock
3. kill nginx, start nginx, send traffic, still use the old lock.

### CODE REVIEW

File: apache2/apache2_config.c [hunk 1]
```diff
@@ -1213,7 +1214,6 @@ static const char *cmd_waf_lock_owner(cmd_parms *cmd,
 
     return NULL;
 }
-#endif
```
Comment (cp5555) [inline apache2/apache2_config.c]: I think it's confused that why you change the WAF_JSON_LOGGING_ENABLE? Maybe you can split one commit for these revision

File: apache2/apache2_util.c [hunk 2]
```diff
@@ -289,24 +290,24 @@ static void get_ruleset_type_version(char* waf_ruleset_info, char* waf_ruleset_t
     }
 }
 
-static int write_file_with_lock(apr_global_mutex_t* lock, apr_file_t* fd, char* str) {
+static int write_file_with_lock(struct waf_lock* lock, apr_file_t* fd, char* str) {
```
Comment (cp5555) [inline apache2/apache2_util.c:293]: Will we rewrite the interface directly?

File: apache2/mod_security2.c [hunk 3]
```diff
@@ -95,9 +95,10 @@ TreeRoot DSOLOCAL *conn_write_state_suspicious_list = 0;
 #ifdef WAF_JSON_LOGGING_ENABLE
 char DSOLOCAL *msc_waf_resourceId = "";
 char DSOLOCAL *msc_waf_instanceId = "";
-char DSOLOCAL *msc_waf_lock_owner = "root";
 #endif
 
+char DSOLOCAL *msc_waf_lock_owner = "root";
```
Comment (cp5555) [inline apache2/mod_security2.c]: I don't know whether it is correct or not because I think the owner should get from nginx option. There are potential bugs if the default owner is "root". According to our discussion, we need to get the user information from nginx configuration.

File: apache2/mod_security2.c [hunk 4]
```diff
@@ -833,7 +834,7 @@ static int hook_post_config(apr_pool_t *mp, apr_pool_t *mp_log, apr_pool_t *mp_t
  * Initialisation performed for every new child process.
  */
 static void hook_child_init(apr_pool_t *mp, server_rec *s) {
-    modsecurity_child_init(modsecurity);
+    modsecurity_child_init(mp, modsecurity);
```
Comment (cp5555) [inline apache2/mod_security2.c]: Could we not change the API?

File: apache2/modsecurity.c [hunk 5]
```diff
@@ -135,13 +135,29 @@ msc_engine *modsecurity_create(apr_pool_t *mp, int processing_mode) {
     return msce;
 }
 
+static void set_lock_args(struct waf_lock_args *lock_args, char *lock_name, int lock_id, char *lock_owner) {
```
Comment (cp5555) [inline apache2/modsecurity.c]: Is it a general function? If so, could we move to the waf lock utility?

File: apache2/modsecurity.c [hunk 6]
```diff
@@ -153,85 +169,53 @@ int modsecurity_init(msc_engine *msce, apr_pool_t *mp) {
 #ifdef WITH_CURL
     curl_global_init(CURL_GLOBAL_ALL);
 #endif
+    lock_args = apr_pcalloc(mp, sizeof(struct waf_lock_args));
+
     /* Serial audit log mutext */
-    rc = apr_global_mutex_create(&msce->auditlog_lock, NULL, APR_LOCK_DEFAULT, mp);
-    if (rc != APR_SUCCESS) {
+    lock_name = apr_pstrcat(mp, "Global\\", "auditlog_lock", NULL);
```
Comment (cp5555) [inline apache2/modsecurity.c]: It may be useless if you just run on Linux.

File: apache2/modsecurity.c [hunk 7]
```diff
@@ -153,85 +169,53 @@ int modsecurity_init(msc_engine *msce, apr_pool_t *mp) {
...
+    rc = Waf_createLock(msce->wafjsonlog_lock, lock_args);    
+    if (Waf_lock_isError(rc)) {
         //ap_log_error(APLOG_MARK, APLOG_ERR, rv, s, "mod_security: Could not create modsec_wafjsonlog_lock");
         //return HTTP_INTERNAL_SERVER_ERROR;
         return -1;
     }
 #endif
 
 #if !defined(MSC_TEST)
-#ifdef __SET_MUTEX_PERMS
```
Comment (cp5555) [inline apache2/modsecurity.c]: Could you add comments for why should remove the code?

File: standalone/server.c [hunk 8]
```diff
@@ -998,26 +998,14 @@ AP_DECLARE(void) unixd_pre_config(apr_pool_t *ptemp)
     apr_finfo_t wrapper;
 
 #if AP_SERVER_MAJORVERSION_NUMBER > 1 && AP_SERVER_MINORVERSION_NUMBER < 3
-#ifdef WAF_JSON_LOGGING_ENABLE
```
Comment (cp5555) [inline standalone/server.c:1001]: Could you add a new commit for WAF_JSON_LOGGING_ENABLE

File: configure.ac [hunk 9]
```diff
@@ -603,7 +603,7 @@ AC_ARG_ENABLE(waf_json_logging,
 [
```
Comment (cp5555) [inline configure.ac:603]: Could you split another commit for BUG FIX?

File: apache2/persist_dbm.c [hunk 10]
```diff
@@ -876,7 +877,7 @@ static int collection_store_ex(int db_option, modsec_rec *msr, apr_table_t *col)
         }
 #ifdef GLOBAL_COLLECTION_LOCK
         apr_sdbm_close(apr_dbm);
-        apr_global_mutex_unlock(msr->modsecurity->dbm_lock);
+	Waf_freeExclusiveLock(msr->modsecurity->dbm_lock);
```
Comment (cp5555) [inline apache2/persist_dbm.c]: alignment issue

File: apache2/persist_dbm.c [hunk 11]
```diff
@@ -865,7 +866,7 @@ static int collection_store_ex(int db_option, modsec_rec *msr, apr_table_t *col)
             if (apr_dbm != NULL) {
 #ifdef GLOBAL_COLLECTION_LOCK
                 apr_sdbm_close(apr_dbm);
-                apr_global_mutex_unlock(msr->modsecurity->dbm_lock);
+	        Waf_freeExclusiveLock(msr->modsecurity->dbm_lock);
```
Comment (cp5555) [inline apache2/persist_dbm.c]: alignment issue

File: apache2/msc_logging.c [hunk 12]
```diff
@@ -2226,8 +2227,8 @@ void sec_audit_logger_native(modsec_rec *msr) {
         sec_auditlog_write(msr, "\n", 1);
 
         /* Unlock the mutex we used to serialise access to the audit log file. */
-        rc = apr_global_mutex_unlock(msr->modsecurity->auditlog_lock);
-        if (rc != APR_SUCCESS) {
+	rc = Waf_freeExclusiveLock(msr->modsecurity->auditlog_lock);
```
Comment (cp5555) [inline apache2/msc_logging.c]: alignment issue

File: apache2/apache2_config.c (hunk 1 above)
Comment (azhao155) [reply apache2/apache2_config.c]: since i am goona use the nginx api to figure out the user, so this is no longer needed

File: apache2/apache2_util.c (hunk 2 above)
Comment (azhao155) [reply apache2/apache2_util.c:293]: yes, i think it's ok

File: apache2/mod_security2.c (hunk 3 above)
Comment (azhao155) [reply apache2/mod_security2.c]: will do that

File: apache2/mod_security2.c (hunk 4 above)
Comment (azhao155) [reply apache2/mod_security2.c]: yes, will use other pool

File: apache2/modsecurity.c (hunk 5 above)
Comment (azhao155) [reply apache2/modsecurity.c]: will make it modsec specific

File: apache2/modsecurity.c (hunk 6 above)
Comment (azhao155) [reply apache2/modsecurity.c]: will make modsec specific

File: apache2/modsecurity.c (hunk 7 above)
Comment (azhao155) [reply apache2/modsecurity.c]: Sure

File: standalone/server.c (hunk 8 above)
Comment (azhao155) [reply standalone/server.c:1001]: Sure

File: configure.ac (hunk 9 above)
Comment (azhao155) [reply configure.ac:603]: Sure

File: apache2/apache2_util.c [hunk 13]
```diff
@@ -289,24 +290,24 @@ static void get_ruleset_type_version(char* waf_ruleset_info, char* waf_ruleset_t
...
+    if (Waf_lock_isError(rc)) {
         return WAF_LOG_UTIL_FAILED;
     }
 
     nbytes = strlen(str);
     apr_file_write_full(fd, str, nbytes, &nbytes_written);
 
-    rc = apr_global_mutex_unlock(lock);
-    if (rc != APR_SUCCESS) {
+    rc = Waf_freeExclusiveLock(lock);
```
Comment (asridharan) [inline apache2/apache2_util.c]: Is this method naming convention idiomatic to Modsecurity? Here we seem to be using a mix of snake and camel case:
https://en.wikipedia.org/wiki/Snake_case

I think we should stick to snake case since that is what seems to be used at other places?

**Review by asridharan — COMMENTED** [review]
@azhao155 could you rebase this PR, to the latest. The Travis CI is comitted, so it will allow us to run the Travis-CI on this PR.

File: apache2/apache2_util.c (hunk 13 above)
Comment (azhao155) [reply apache2/apache2_util.c]: will update all the format in the external api

**Review by asridharan — APPROVED** [review]
